### PR TITLE
New ATC Tables

### PR DESCRIPTION
**Pull request by obelisk**
Tl;dr: You can expose SQLite database queries as osquery tables.

This creates a new concept of what I've been calling ATC table (Auto Table Construction). The concept is that you can provide a path to an SQLite database, a query to run against it, and the columns you want and osquery will construct a table that pulls that information for you.

I built this to prevent the code pattern we were starting to see which consisted of opening a database, doing a query, copying the data over, and all the code duplication that came with that.

For example, quarantine_items can be expressed (mostly) as a new configuration block:
```JSON
    "auto_table_construction" : {
        "quarantine_items" : {
          "query" : "SELECT LSQuarantineEventIdentifier as id, LSQuarantineAgentName as agent_name, LSQuarantineAgentBundleIdentifier as agent_bundle_identifier, LSQuarantineTypeNumber as type, LSQuarantineDataURLString as data_url,LSQuarantineOriginURLString as origin_url, LSQuarantineSenderName as sender_name, LSQuarantineSenderAddress as sender_address, LSQuarantineTimeStamp as timestamp from LSQuarantineEvent",
          "path" : "/Users/%/Library/Preferences/com.apple.LaunchServices.QuarantineEventsV2",
          "columns" : ["path", "id", "agent_name", "agent_bundle_identifier"]
        }
    }
```
(Shoutout to @terracatta for the query and idea for this ATC example)

There is still progress to be made:

- [x] Make paths support globbing
- [x] Check there is no registration over existing tables
- [x] Test it more
- [x] Write tests
- [x] ATC Platform Specification 

But I'm putting the code up now because the base functionality works and I would like to solicit feedback from the community. There is a new API for this as well: `genQueryDataForSqliteTable` if people want to use it.

More notable changes:
* If a configuration file does not contain a key for a config parser, an empty one is no longer create. The only config parser that seemed to have issues with this was YARA.
* The yara code was changed slightly to accommodate the lack of YARA keys.

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit 9cf1a2df03ef5adb83f81b1cfa5423331154449a (Job results: [2843](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2843/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 9cf1a2df03ef5adb83f81b1cfa5423331154449a (Job results: [3711](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3711/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 9cf1a2df03ef5adb83f81b1cfa5423331154449a (Job results: [6819](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6819/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 9cf1a2df03ef5adb83f81b1cfa5423331154449a (Job results: [3712](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3712/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 9cf1a2df03ef5adb83f81b1cfa5423331154449a (Job results: [2844](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2844/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 9cf1a2df03ef5adb83f81b1cfa5423331154449a (Job results: [6820](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6820/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 0280eec143b48257f86394d3fec658dcced2dc0e (Job results: [3713](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3713/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 0280eec143b48257f86394d3fec658dcced2dc0e (Job results: [3714](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3714/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit f83bb8398c5d51254c9e23eb37d83c006c11e4c4 (Job results: [3715](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3715/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit f83bb8398c5d51254c9e23eb37d83c006c11e4c4 (Job results: [3716](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3716/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:no_good: The commit 130e70f72c7132e9042297f9afd19bdb30ebe98a (Job results: [3328](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3328/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 130e70f72c7132e9042297f9afd19bdb30ebe98a (Job results: [3717](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3717/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 66350e0cfb4e48432f6f86bb3ed0bc222381cf07 (Job results: [3718](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3718/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 1d1e97b6b001bb083e28ee6acdfeeb5eb94308ea (Job results: [2855](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2855/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:no_good: The commit 1d1e97b6b001bb083e28ee6acdfeeb5eb94308ea (Job results: [3335](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3335/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 677ecbbc7615b47166b9ec5af740eafc2d99c546 (Job results: [2857](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2857/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 677ecbbc7615b47166b9ec5af740eafc2d99c546 (Job results: [3728](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3728/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 677ecbbc7615b47166b9ec5af740eafc2d99c546 (Job results: [6831](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6831/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 677ecbbc7615b47166b9ec5af740eafc2d99c546 (Job results: [2859](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2859/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 677ecbbc7615b47166b9ec5af740eafc2d99c546 (Job results: [6833](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6833/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit f11412e5002c8c04dd5a3ad9d316ff7a2a10d206 (Job results: [2861](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2861/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit f11412e5002c8c04dd5a3ad9d316ff7a2a10d206 (Job results: [3732](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3732/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit f11412e5002c8c04dd5a3ad9d316ff7a2a10d206 (Job results: [2863](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2863/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit f11412e5002c8c04dd5a3ad9d316ff7a2a10d206 (Job results: [6837](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6837/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:no_good: The commit 72a8238bf5954e7705231ba0f1d54c3f1e7b9ced (Job results: [3341](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3341/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 72a8238bf5954e7705231ba0f1d54c3f1e7b9ced (Job results: [6838](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6838/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 72a8238bf5954e7705231ba0f1d54c3f1e7b9ced (Job results: [2864](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2864/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 72a8238bf5954e7705231ba0f1d54c3f1e7b9ced (Job results: [3735](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3735/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 72a8238bf5954e7705231ba0f1d54c3f1e7b9ced (Job results: [2865](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2865/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 72a8238bf5954e7705231ba0f1d54c3f1e7b9ced (Job results: [6839](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6839/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:no_good: The commit 7af7bf0565f94da9e18161bb9845ae3595304a40 (Job results: [3342](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3342/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 2dc97bc08c72aa55aa56f410e47766abda30058f (Job results: [3343](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3343/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 94fdc9fba60ca436f97d50d88f9e8ec668298c87 (Job results: [3739](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3739/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 94fdc9fba60ca436f97d50d88f9e8ec668298c87 (Job results: [6842](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6842/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 94fdc9fba60ca436f97d50d88f9e8ec668298c87 (Job results: [2869](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2869/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 94fdc9fba60ca436f97d50d88f9e8ec668298c87 (Job results: [3740](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3740/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 94fdc9fba60ca436f97d50d88f9e8ec668298c87 (Job results: [6843](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6843/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:no_good: The commit 940afab61a700573800caaf8b43f6b7be664ad9f (Job results: [3346](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3346/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 715bc715703f1ca08920f80d230f5b470d778e32 (Job results: [6953](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6953/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 715bc715703f1ca08920f80d230f5b470d778e32 (Job results: [1389](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1389/)) failed one or more tests (**FreeBSD**).

File: osquery/config/additional_parsers/auto_constructed_tables.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+  }
+
+  return (platform.find(osquery::kSDKPlatform) != std::string::npos);
+}
+
+/**
+ * @brief A ConfigParserPlugin for ATC (Auto Table Construction)
+ */
+class ATCConfigParserPlugin : public ConfigParserPlugin {
+  const std::string kParserKey = "auto_table_construction";
```
Comment (muffins) [inline osquery/config/additional_parsers/auto_constructed_tables.cpp]: nit: Any reason for non-bracket initialization?

File: osquery/config/additional_parsers/auto_constructed_tables.cpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                          : ""};
+    std::string path{params.HasMember("path") && params["path"].IsString()
+                         ? params["path"].GetString()
+                         : ""};
+    std::string platform{params.HasMember("platform") &&
+                                 params["platform"].IsString()
+                             ? params["platform"].GetString()
+                             : ""};
+    
+    if (query == "" || path == "") {
```
Comment (muffins) [inline osquery/config/additional_parsers/auto_constructed_tables.cpp]: prefer `.empty()`

File: osquery/config/additional_parsers/auto_constructed_tables.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    setDatabaseValue(
+        kPersistentSettings, kDatabaseKeyPrefix + table_name, table_settings);
+    s = tables->add(
+        table_name, std::make_shared<ATCPlugin>(path, columns, query), true);
+
+    PluginResponse resp;
+    Registry::call(
+        "sql", "sql", {{"action", "attach"}, {"table", table_name}}, resp);
+
+    if (!s.ok()) {
```
Comment (muffins) [inline osquery/config/additional_parsers/auto_constructed_tables.cpp]: let's move this up above the `Registry::call` in case the `tables->add` fails.

File: osquery/config/config.cpp [hunk 4]
```diff
@@ -666,8 +666,6 @@ void Config::applyParsers(const std::string& source,
 
         auto doc = JSON::newFromValue(obj[key]);
         parser_config.emplace(std::make_pair(key, std::move(doc)));
-      } else {
-        parser_config.emplace(std::make_pair(key, JSON::newObject()));
```
Comment (muffins) [inline osquery/config/config.cpp]: Why is this being removed? I'm not sure I follow.

File: osquery/sql/virtual_sqlite_table.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+namespace fs = boost::filesystem;
+
+namespace osquery {
+
+Status genSqliteQueryRow(sqlite3_stmt* stmt,
+                         QueryData& qd,
+                         const fs::path& sqlite_db) {
+  Row r;
+  for (auto i{0}; i < sqlite3_column_count(stmt); ++i) {
```
Comment (muffins) [inline osquery/sql/virtual_sqlite_table.cpp]: nit: my own personal preference here, but I feel like the braced initialization is a bit excessive. Why not leave the typing and equal here as a normal `for` loop? Am I coming off as too old and fighting the future? :)

File: osquery/sql/virtual_sqlite_table.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    auto column_name = std::string(sqlite3_column_name(stmt, i));
+    auto column_type = sqlite3_column_type(stmt, i);
+    switch (column_type) {
+    case SQLITE_TEXT: {
+      auto text_value = sqlite3_column_text(stmt, i);
+      if (text_value != nullptr) {
+        r[column_name] = std::string(reinterpret_cast<const char*>(text_value));
```
Comment (muffins) [inline osquery/sql/virtual_sqlite_table.cpp]: are you sure there's truly no way to avoid this `reinterpret_cast`?

File: osquery/sql/virtual_sqlite_table.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    }
+  }
+  if (r.count("path") > 0) {
+    LOG(WARNING) << "Row contains a path key, refusing to overwrite";
+  } else {
+    r["path"] = sqlite_db.string();
+  }
+  qd.push_back(r);
+  return Status{};
```
Comment (muffins) [inline osquery/sql/virtual_sqlite_table.cpp]: nit: Why braced? Why not `Status()` as we have it in the rest of the code base?

File: tools/tests/test_base.py [hunk 8]
```diff
@@ -150,6 +150,7 @@ def __init__(self, command='../osqueryi', args={}, env={}):
         options["database_path"] += str(random.randint(1000, 9999))
         command = command + " " + " ".join(
             ["--%s=%s" % (k, v) for k, v in options.iteritems()])
+        print(command)
```
Comment (muffins) [inline tools/tests/test_base.py]: Is this needed? I'd say remove this for production, or add some additional flair to make it clear it's for debugging purposes.

File: osquery/tables/yara/yara_utils.cpp [hunk 9]
```diff
@@ -260,6 +260,9 @@ Status YARAConfigParserPlugin::setUp() {
 Status YARAConfigParserPlugin::update(const std::string& source,
                                       const ParserConfig& config) {
   // The YARA config parser requested the "yara" top-level key in the config.
+  if (config.count("yara") == 0) {
+    return Status(1, "No Yara Configuration");
```
Comment (muffins) [inline osquery/tables/yara/yara_utils.cpp]: Should this be a `Status(1...`? I'm under the impression it's not an _error_ if the yara config doesn't load, but I'm also writing similar logic for this and want to be consistent.

File: osquery/config/additional_parsers/auto_constructed_tables.cpp (hunk 1 above)
Comment (obelisk) [reply osquery/config/additional_parsers/auto_constructed_tables.cpp]: omg I have shamed my family.

File: osquery/config/config.cpp (hunk 4 above)
Comment (obelisk) [reply osquery/config/config.cpp]: This is being removed because if you are using multiple config files, you get an empty object from all files that do not contain the a block for your config.

In ATC's case, this makes it impossible to tell if the ATC configuration was removed, or this is just another file that does not contain an ATC block.

The expected configuration for ATC should be an ATC block in one config file, so it knows what to remove when it changes. If an ATC block is removed entirely, it keeps tables, if the ATC block is emptied (left as an empty object) it removes all ATC tables.

File: osquery/sql/virtual_sqlite_table.cpp (hunk 5 above)
Comment (obelisk) [reply osquery/sql/virtual_sqlite_table.cpp]: I mean technically I think this should be: `decltype(sqlite3_column_count(stmt)) i{0};`

File: osquery/sql/virtual_sqlite_table.cpp (hunk 6 above)
Comment (obelisk) [reply osquery/sql/virtual_sqlite_table.cpp]: If we want it a const char* then I think so but it's type without the cast is `const unsigned char *` which may be sufficient.

File: osquery/tables/yara/yara_utils.cpp (hunk 9 above)
Comment (obelisk) [reply osquery/tables/yara/yara_utils.cpp]: This is a side effect of the the JSON object removal above. I would agree though this is not an error.

File: osquery/config/additional_parsers/auto_constructed_tables.cpp [hunk 10]
```diff
@@ -0,0 +1,193 @@
+/**
```
[the remark this answers is not in the document]
Comment (obelisk) [reply osquery/config/additional_parsers/auto_constructed_tables.cpp]: Done

**Comment by osqueryer** [comment]
:no_good: The commit 6f12d9fad418be147887cac3875ef42b5acf53c2 (Job results: [3451](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3451/)) failed the code audit or documentation test.

File: osquery/sql/virtual_sqlite_table.cpp (hunk 6 above)
Comment (obelisk) [reply osquery/sql/virtual_sqlite_table.cpp]: Turns out it is not sufficient and a reinterpret cast is needed.

**Comment by osqueryer** [comment]
:-1: The commit 6f12d9fad418be147887cac3875ef42b5acf53c2 (Job results: [2986](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2986/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:no_good: The commit 14ae40e69b8fa8ca2a0cbd4820dc42a268705f5d (Job results: [3454](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3454/)) failed the code audit or documentation test.

File: osquery/config/additional_parsers/auto_constructed_tables.cpp [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+  }
+  return qd;
+}
+
+/// Remove these ATC tables from the registry and database
+Status ATCConfigParserPlugin::removeATCTables(
+    const std::set<std::string>& detach_tables) {
+  auto registry_table = RegistryFactory::get().registry("table");
+  for (const auto& table : detach_tables) {
+    if (registry_table->exists(table)) {
```
Comment (muffins) [inline osquery/config/additional_parsers/auto_constructed_tables.cpp]: A couple of questions about this block. 
* You make a check for `exists`, and exit completely if _one_ table doesn't exist, why not `LOG(WARNING)` and `continue`?
* Continuing on that last point, could you flip the logic and `continue` early sans `else` clause? This might save you a level of indentation, and could contribute to the next point
* What if a table exists, but the `getDatabaseValue` fails? Perhaps you could bring the `getDatabaseValue` out of the `if`, and should it fail emit a `LOG(WARNING)` that the DB lookup failed?

File: osquery/config/additional_parsers/auto_constructed_tables.cpp [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+  }
+  return set_tables;
+}
+
+Status ATCConfigParserPlugin::setUp() {
+  VLOG(1) << "Removing stale ATC entries";
+  std::vector<std::string> keys;
+  scanDatabaseKeys(kPersistentSettings, keys, kDatabaseKeyPrefix);
+  for (const auto& key : keys) {
+    deleteDatabaseValue(kPersistentSettings, key);
```
Comment (muffins) [inline osquery/config/additional_parsers/auto_constructed_tables.cpp]: What about capturing the output of this and checking?
```
auto s = deleteDatabaseValue(kPersistentSettings, key);
if (!s.ok()) {
  VLOG(1) << "Setup failed, things bad, error, system fail, sad times :(";
}
...
```

File: osquery/config/additional_parsers/auto_constructed_tables.cpp [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+      continue;
+    }
+
+    // Remove the old table to replace with the new one
+    s = removeATCTables({table_name});
+    if (!s.ok()) {
+      LOG(WARNING) << "ATC table overrides core table; Refusing registration";
+      continue;
+    }
+    setDatabaseValue(
```
Comment (muffins) [inline osquery/config/additional_parsers/auto_constructed_tables.cpp]: Maybe check the output here? As your flow seems to rely on it?

File: osquery/config/additional_parsers/auto_constructed_tables.cpp (hunk 11 above)
Comment (obelisk) [reply osquery/config/additional_parsers/auto_constructed_tables.cpp]: 1) Good idea
2) We don't want to flip it because we want the side effect of removing from the db.
3) If the table exists but get fails then the table is a core table (registered but not in the db) so that's the message we create
